### PR TITLE
ADR-144 Realm resolution from connection string

### LINKED ISSUE #144: Realm resolution from connection string
---
layout: adr
adr: 144
title: Realm resolution from connection string
date: 2022-12-13
status: Review
type: Standards Track
spdx-license: CC0-1.0
authors:
  - menduz
---

## Abstract

This document describes the Realm resolution algorithms to consistently resolve servers based on a realm connection string. It resembles DNS, and uses the DAO catalysts, ENS and URLs to resolve the realms as described in [ADR-110](https://adr.decentraland.org/adr/ADR-110).

## Context, Reach & Prioritization

Realms ([ADR-110](https://adr.decentraland.org/adr/ADR-110)) are the configurations that define what and how is executed and rendered in a Decentraland Explorer [ADR-102](https://adr.decentraland.org/adr/ADR-102). If we reason by analogy and consider an Explorer like a Web Browser, and the Scenes like WebSites. Then the Realm is the URL of that site.

## Specification

Realm are represented as strings. Those strings need to be resolved to a `/about` endpoint to fetch the configuration.

The mechanism to resolve that endpoint is:
1. If the string is an URL, then it uses the URL
2. If the string equals to the name of a DAO catalyst, then it picks the URL of the Catalyst like `https://peer.decentraland.org/about`
3. If the string is an ENS, then it resolves using ENS `dcl.realm` TXT values
   1. If no values are defined for the ENS-DNS and it is a `.dcl.eth` subdomain, then it falls back to the Worlds resolver
4. If this point is reached, then the realm is invalid.

### ENS resolver

NAME and ENS holders can decide where their worlds are hosted. For that purpose, a special record `dcl.realm` can be added both to ENS and NAME tokens and it should resolve to a base URL that is compliant with the [Realm specification (ADR-110)](https://adr.decentraland.org/adr/ADR-110)

### Worlds resolver

Worlds is a configuration of services that enable deploying Decentraland scenes using NAME, as described in [ADR-111](/adr/ADR-111). To resolve Worlds, an URL in the shape of `https://worlds-content-server.decentraland.org/world/<NAME>.dcl.eth` is used, the `/about` response is compliant with the [Realm specification (ADR-110)](https://adr.decentraland.org/adr/ADR-110)

## RFC 2119 and RFC 8174

> The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

### PR DESCRIPTION
**Pull request by menduz**
closes #144

Signed-off-by: menduz <github@menduz.com>

### CODE REVIEW

File: content/ADR-144.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
+---
+layout: adr
+adr: 144
```
Comment (marianogoldman) [inline content/ADR-144.md:3]:
```suggestion
adr: 150
```
And also rename the file

File: content/ADR-144.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+<!--
+Discuss and go into detail about the subject in question. Make sure you cover:
+- Why is this decision important
+- The urgency of the decision
+- Datapoints and related background information
+- Vocabulary and key terms
+-->
+
+Realms ([ADR-110](https://adr.decentraland.org/adr/ADR-110)) are the configurations that define what and how is executed and rendered in a Decentraland Explorer [ADR-102](https://adr.decentraland.org/adr/ADR-102). If we reason by analogy and consider an Explorer like a Web Browser, and the Scenes like WebSites. Then the Realm is the URL of that site.
```
Comment (marianogoldman) [inline content/ADR-144.md:31]:
```suggestion
Realms ([ADR-110](https://adr.decentraland.org/adr/ADR-110)) are the configurations that define what and how is executed and rendered in a Decentraland Explorer ([ADR-102](https://adr.decentraland.org/adr/ADR-102)). If we reason by analogy and consider an Explorer like a Web Browser, and the Scenes like WebSites. Then the Realm is the URL of that site.
```

File: content/ADR-144.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+## Specification
+
+<!--
+The technical specification should describe the syntax and semantics of any new feature.
+-->
+
+Realm are represented as strings. Those strings need to be resolved to a `/about` endpoint to fetch the configuration.
```
Comment (marianogoldman) [inline content/ADR-144.md:39]:
```suggestion
Realms are represented as strings. Those strings need to be resolved to an `/about` endpoint to fetch the configuration.
```

File: content/ADR-144.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+
+The mechanism to resolve that endpoint is:
+1. If the string is an URL, then it uses the URL
```
Comment (marianogoldman) [inline content/ADR-144.md:43]: Shouldn't it add `/about` at the end? For e.g. for worlds the URL should be `https://my-worlds-server.io/world/my-name.dcl.eth`, not `https://my-worlds-server.io/world/my-name.dcl.eth/about`

File: content/ADR-144.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+   1. If no values are defined for the ENS-DNS and it is a `.dcl.eth` subdomain, then it falls back to the Worlds resolver
+4. If this point is reached, then the realm is invalid.
+
+### ENS resolver
+
+NAME and ENS holders can decide where their worlds are hosted. For that purpose, a special record `dcl.realm` can be added both to ENS and NAME tokens and it should resolve to a base URL that is compliant with the [Realm specification (ADR-110)](https://adr.decentraland.org/adr/ADR-110)
+
+### Worlds resolver
+
+Worlds is a configuration of services that enable deploying Decentraland scenes using NAME, as described in [ADR-111](/adr/ADR-111). To resolve Worlds, an URL in the shape of `https://worlds-content-server.decentraland.org/world/<NAME>.dcl.eth` is used, the `/about` response is compliant with the [Realm specification (ADR-110)](https://adr.decentraland.org/adr/ADR-110)
```
Comment (marianogoldman) [inline content/ADR-144.md:55]:
```suggestion
Worlds is a configuration of services that enable deploying Decentraland scenes using a NAME, as described in [ADR-111](/adr/ADR-111). To resolve Worlds, a URL in the shape of `https://worlds-content-server.decentraland.org/world/<NAME>.dcl.eth` is used, the `/about` response is compliant with the [Realm specification (ADR-110)](https://adr.decentraland.org/adr/ADR-110)
```

File: content/ADR-144.md (hunk 1 above)
Comment (menduz) [reply content/ADR-144.md:3]: The issue was created to park the ID 144

**Review by marianogoldman — APPROVED** [review]

**Comment by pentreathm** [comment]
amazing!

**Pull request merged**